src/openslide-vendor-philips-isyntax.c

```diff
@@ -0,0 +1,397 @@
+    for (int y = start; y < end; ++y) {
+        tile_pixels[y*tile_width + x] = color;
+    }
+}
+
+static void draw_text(uint32_t* tile_pixels, int32_t tile_width, int32_t x_pos, int32_t y_pos, uint32_t color, const char* text) {
+#if IS_DEBUG_ANNOTATE_TILE
+    const int font_size = 8;
+    for (const char* ch = text; *ch != 0; ++ch) {
+        for (int y = 0; y < font_size; ++y) {
+            uint8_t bit_line = font8x8_basic[*((u8*)ch)][y];
+            for (int x = 0; x < font_size; ++x) {
+                if (bit_line & (1u << x)) {
+                    tile_pixels[(y + y_pos) * tile_width + x + x_pos] = color;
+                }
+            }
+        }
+        x_pos += font_size;
+    }
+#else
+    // Unused variable warning suppression.
+    (void)tile_pixels;
+    (void)tile_width;
+    (void)x_pos;
+    (void)y_pos;
+    (void)color;
+    (void)text;
+#endif
+}
+
+static void annotate_tile(uint32_t* tile_pixels, int32_t scale, int32_t tile_col, int32_t tile_row, int32_t tile_width, int32_t tile_height) {
+    // TODO(avirodov): maybe move this to libisyntax.
+    if (IS_DEBUG_ANNOTATE_TILE) {
+        // OpenCV in C is hard... the core_c.h includes types_c.h which includes cvdef.h which is c++.
+        // But we don't need much. Axis-aligned lines, and some simple text.
+        int pad = 1;
+        uint32_t color = 0xff0000ff; // ARGB
+        draw_horiz_line(tile_pixels, tile_width, /*y=*/pad, /*start=*/pad, /*end=*/tile_width - pad, color);
+        draw_horiz_line(tile_pixels, tile_width, /*y=*/tile_height - pad, /*start=*/pad, /*end=*/tile_width - pad, color);
+
+        draw_vert_line(tile_pixels, tile_width, /*x=*/pad, /*start=*/pad, /*end=*/tile_height - pad, color);
+        draw_vert_line(tile_pixels, tile_width, /*x=*/tile_width - pad, /*start=*/pad, /*end=*/tile_height - pad, color);
+
+        char buf[128];
+        sprintf(buf, "x=%d,y=%d,s=%d", tile_row, tile_col, scale);
+        draw_text(tile_pixels, tile_width, 10, 10, color, buf);
+    }
+}
+
+static bool philips_isyntax_detect(
+        const char *filename,
+        struct _openslide_tifflike *tl G_GNUC_UNUSED,
+        GError **err G_GNUC_UNUSED) {
+    LOG("got filename %s", filename);
+    LOG_VAR("%p", tl);
+    // reject TIFFs
+    if (tl) {
+        g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED,
+                    "Is a TIFF file");
+        return false;
+    }
+
+    g_autoptr(_openslide_file) f = _openslide_fopen(filename, err);
+    if (f == NULL) {
+        LOG("Failed to open file");
+        return false;
+    }
+
+    const int num_chars_to_read = 256;
+    g_autofree char *buf = g_malloc(num_chars_to_read);
+    size_t num_read = _openslide_fread(f, buf, num_chars_to_read-1);
+    buf[num_chars_to_read-1] = 0;
+    LOG_VAR("%ld", num_read);
+    LOG_VAR("%s", buf);
+
+    // TODO(avirodov): probably a more robust XML parsing is needed.
```

Comment:
> But would it make sense to just open the file with libisyntax, or is that too expensive?

> Proposal for this CL: leave as-is with TODO, open bug in libisyntax to have a lightweight 'detect' function. I will wait on opening the bug until I hear back here in case there are other suggestions.

It sounds like a good idea to me to add a lightweight detect function to libisyntax.
I agree that a strstr() for the first tag is probably all we need in order to fail at the earliest possible time, weeding out practically all non-iSyntax files. Parsing the whole XML header as we do in libisyntax_open() is indeed expensive, as it can be up to hundreds of megabytes, also containing large sections of embedded base64, with the size of the XML header initially unknown.